### ISSUE TITLE
rpull pulls all layers?

### ISSUE THREAD

**Issue by nicktming** [containerd/stargz-snapshotter#254]
containerd: v1.4.3
stargz-snapshotter: v0.3.0
# ctr-remote images rpull --plain-http 49.232.118.181:5000/myimage:v0.1
fetching sha256:d737a6fd... application/vnd.docker.distribution.manifest.v2+json
fetching sha256:8c4e826b... application/vnd.docker.container.image.v1+json
fetching sha256:0703b966... application/vnd.docker.image.rootfs.diff.tar.gzip
fetching sha256:e966595c... application/vnd.docker.image.rootfs.diff.tar.gzip
fetching sha256:3b36b637... application/vnd.docker.image.rootfs.diff.tar.gzip

is it normal case?  it pulls the tar.gzip besides image manifest/config files. and it is pending here, rpull takes a long time.

**Comment by ktock** [comment]
Have you done registry configuration on the snapshotter's side?

https://github.com/containerd/stargz-snapshotter/blob/v0.3.0/docs/overview.md#registry-mirrors-and-insecure-connection

something like (/etc/containerd-stargz-grpc/config.toml):

```toml
[[resolver.host."49.232.118.181:5000".mirrors]]
host = "49.232.118.181:5000"
insecure = true
```

If it doesn't work, please provide the debug log of `containerd-stargz-grpc`.

**Comment by nicktming** [comment]
> Have you done registry configuration on the snapshotter's side?
> 
> https://github.com/containerd/stargz-snapshotter/blob/v0.3.0/docs/overview.md#registry-mirrors-and-insecure-connection
> 
> something like (/etc/containerd-stargz-grpc/config.toml):
> 
> ```toml
> [[resolver.host."49.232.118.181:5000".mirrors]]
> host = "49.232.118.181:5000"
> insecure = true
> ```
> 
> If it doesn't work, please provide the debug log of `containerd-stargz-grpc`.

it still pulls the tar.gzip layers. by the way, i forget to hide my ip address, would you mind hiding it in your comment?

**Comment by ktock** [comment]
@nicktming Could you provide the log output from `containerd-stargz-grpc`?

**Comment by nicktming** [comment]
> @nicktming Could you provide the log output from `containerd-stargz-grpc`?

there is no output
my containerd config
`
# Plug stargz snapshotter into containerd
# Containerd recognizes stargz snapshotter through specified socket address.
# The specified address below is the default which stargz snapshotter listen to.
[proxy_plugins]
  [proxy_plugins.stargz]
    type = "snapshot"
    address = "/run/containerd-stargz-grpc/containerd-stargz-grpc.sock"

# Use stargz snapshotter through CRI
[plugins."io.containerd.grpc.v1.cri".containerd]
  snapshotter = "stargz"
  disable_snapshot_annotations = false
`
i already covert my ubuntu image to stargz-format image and pushed to my private registry.

**Comment by nicktming** [comment]
another problem:

1. started `containerd-stargz-grpc -log-level=debug`
2. started `containerd`
3. got error: snapshotter not loaded: stargz: invalid argument
```
ctr-remote images rpull docker.io/stargz/golang:1.12.9-esgz
fetching sha256:e31d9d69... application/vnd.docker.distribution.manifest.v2+json
fetching sha256:7a26855a... application/vnd.docker.container.image.v1+json
ctr: failed to stat snapshot sha256:58461afcf82ffebf2f3f898f5bfd3f4fb6210dd32f21a9145c659c382938ea9e: snapshotter not loaded: stargz: invalid argument
```
anything missed?

**Comment by nicktming** [comment]
i tested the `docker.io/stargz/golang:1.12.9-esgz`
```
fetching sha256:e31d9d69... application/vnd.docker.distribution.manifest.v2+json
fetching sha256:7a26855a... application/vnd.docker.container.image.v1+json
fetching sha256:7b640c31... application/vnd.docker.image.rootfs.diff.tar.gzip
fetching sha256:1a2751b9... application/vnd.docker.image.rootfs.diff.tar.gzip
fetching sha256:cc61f9f9... application/vnd.docker.image.rootfs.diff.tar.gzip
fetching sha256:191d7ea8... application/vnd.docker.image.rootfs.diff.tar.gzip
fetching sha256:18c99e16... application/vnd.docker.image.rootfs.diff.tar.gzip
fetching sha256:8ea059e7... application/vnd.docker.image.rootfs.diff.tar.gzip
fetching sha256:3d16b66d... application/vnd.docker.image.rootfs.diff.tar.gzip
```
it pulls all the layers, is it the normal case?

**Comment by ktock** [comment]
@nicktming 

> there is no output

Please add `--log-level=debug` to `containerd-stargz-grpc` and paste the log output.

> it pulls all the layers, is it the normal case?

no.

Check also `fusermount` is installed on your host.

**Comment by nicktming** [comment]
> @nicktming
> 
> > there is no output
> 
> Please add `--log-level=debug` to `containerd-stargz-grpc` and paste the log output.
> 
> > it pulls all the layers, is it the normal case?
> --- already added, still no logs
> 
> no.
> 
> Check also `fusermount` is installed on your host.
--- 
already installed.
[root@VM-0-6-centos ~]# which fuser
/usr/sbin/fuser
[root@VM-0-6-centos ~]# which fusermount 
/usr/bin/fusermount

**Comment by ktock** [comment]
@nicktming Doesn't `containerd-stargz-grpc` provide any log even with `--log-level=debug` flag?

**Comment by nicktming** [comment]
> @nicktming Doesn't `containerd-stargz-grpc` provide any log even with `--log-level=debug` flag?

it has logs now, logs like that:
`
failed to request: context canceled: failed to resolve","level":"debug",
`

**Comment by ktock** [comment]
@nicktming Please provide the complete log line. It's hard to make sure where is the root cause of that error.

**Comment by nicktming** [comment]
> @nicktming Please provide the complete log line. It's hard to make sure where is the root cause of that error.

containerd logs
```
INFO[2021-02-02T17:15:39.912527373+08:00] StopPodSandbox for "01c5ba45767d4740688ddc50796afb9504c2101b2bdd4fe9e9a79a82de4ec241" 
ERRO[2021-02-02T17:15:39.914353327+08:00] StopPodSandbox for "01c5ba45767d4740688ddc50796afb9504c2101b2bdd4fe9e9a79a82de4ec241" failed  error="failed to destroy network for sandbox \"01c5ba45767d4740688ddc50796afb9504c2101b2bdd4fe9e9a79a82de4ec241\": invalid version \"\": the version is empty"
INFO[2021-02-02T17:15:39.914565152+08:00] StopPodSandbox for "1283bc3c639971d5b6257f893e287ff02d6a98567711dc3168abd757e7d1148f" 
ERRO[2021-02-02T17:15:39.916446931+08:00] StopPodSandbox for "1283bc3c639971d5b6257f893e287ff02d6a98567711dc3168abd757e7d1148f" failed  error="failed to destroy network for sandbox \"1283bc3c639971d5b6257f893e287ff02d6a98567711dc3168abd757e7d1148f\": invalid version \"\": the version is empty"
INFO[2021-02-02T17:15:39.916648867+08:00] StopPodSandbox for "2bb915d797896b8f7a3151a6b0b2187b2664b9c4c77180d2af334a4ba1c1c349" 
ERRO[2021-02-02T17:15:39.918794437+08:00] StopPodSandbox for "2bb915d797896b8f7a3151a6b0b2187b2664b9c4c77180d2af334a4ba1c1c349" failed  error="failed to destroy network for sandbox \"2bb915d797896b8f7a3151a6b0b2187b2664b9c4c77180d2af334a4ba1c1c349\": invalid version \"\": the version is empty"
INFO[2021-02-02T17:15:39.919055945+08:00] StopPodSandbox for "3cb633e1b7909a958839dcd6ffbd924280d474010fdb1fc5df78a1d95b498933" 
ERRO[2021-02-02T17:15:39.921034623+08:00] StopPodSandbox for "3cb633e1b7909a958839dcd6ffbd924280d474010fdb1fc5df78a1d95b498933" failed  error="failed to destroy network for sandbox \"3cb633e1b7909a958839dcd6ffbd924280d474010fdb1fc5df78a1d95b498933\": invalid version \"\": the version is empty"
```

containerd-stargz logs
```
{"error":"cannot resolve layer: failed to redirect (host \"ip:5000\", ref:\"ip:5000/myimage\", digest:\"sha256:a8bb704b3737af43af9314f7fc628e1218703fbdde08cb42681c679efe03aee4\"): failed to request: context canceled: failed to resolve","level":"debug","mountpoint":"/var/lib/containerd-stargz-grpc/snapshotter/snapshots/29/fs","msg":"failed to resolve source","src":"ip:5000/myimage/sha256:a8bb704b3737af43af9314f7fc628e1218703fbdde08cb42681c679efe03aee4","time":"2021-02-02T16:35:28.350105129+08:00"}
{"error":"cannot resolve layer: failed to redirect (host \"ip:5000\", ref:\"ip:5000/myimage\", digest:\"sha256:30721a2fe2615f34937b2aa463b8f741102bf450e76578a07e4fd9bc9e865ff3\"): failed to request: context canceled: failed to resolve","level":"debug","mountpoint":"/var/lib/containerd-stargz-grpc/snapshotter/snapshots/29/fs","msg":"failed to resolve source","src":"ip:5000/myimage/sha256:30721a2fe2615f34937b2aa463b8f741102bf450e76578a07e4fd9bc9e865ff3","time":"2021-02-02T16:35:28.350127541+08:00"}
```

**Comment by ktock** [comment]
@nicktming

> {"error":"cannot resolve layer: failed to redirect (host \"ip:5000\", ref:\"ip:5000/myimage", digest:\"sha256:2ddd81e5692f21714879195f97d10cf39da3d448c014a0c66727914f20886656\"): failed to request: tls: first record does not look like a TLS handshake: failed to resolve","level":"debug","mountpoint":"/var/lib/containerd-stargz-grpc/snapshotter/snapshots/31/fs","msg":"failed to resolve source","src":"ip:5000/myimage/sha256:2ddd81e5692f21714879195f97d10cf39da3d448c014a0c66727914f20886656","time":"2021-02-02T17:22:03.989493832+08:00"}

this log says that you are trying to pull the image from your non-HTTPS (i.e. plain-http) registry `ip:5000` but `containerd-stargz-grpc` doesn't recognize it as non-HTTPS registry and this tries to connect to your registry using HTTPS (then it fails).

As mentioned in https://github.com/containerd/stargz-snapshotter/issues/254#issuecomment-771369476, please make sure that your `containerd-stargz-grpc` configuration (tipically /etc/containerd-stargz-grpc/config.toml) contains the following registry configuration and `containerd-stargz-grpc` recognizes that config file correctly ( /etc/containerd-stargz-grpc/config.toml is read by default but if you store this config file somewhere else, specify the location using `containerd-stargz-grpc`'s `--config` option).

https://github.com/containerd/stargz-snapshotter/blob/v0.3.0/docs/overview.md#registry-mirrors-and-insecure-connection

```toml
[[resolver.host."<ip>:5000".mirrors]]
host = "<ip>:5000"
insecure = true
```

(replace `<ip>` with your true registry host name or IP address)

**Comment by nicktming** [comment]
@ktock what is the stability of stargz-snapshotter in the range of different cluster nodes?  such as 100 nodes, 1000+ nodes, 5000+ nodes .  is there any docs to provide this kind of information?

**Comment by ktock** [comment]
Though there is no survey about stability, this is one of the topics we'll work on this year.

**Comment by nicktming** [comment]
@ktock  ok, thx. Looking forward to the survey.

**Comment by ktock** [comment]
@nicktming 
Of course, any contribution around stability is always welcome :)
BTW how is this issue going?

**Comment by nicktming** [comment]
@ktock it worked with your suggestions, but it failed some times after that with the same command, so i got a little confused.

i will get into the details to find what happens if having free times .

**Comment by mc256** [comment]
**TL;DR**
You may need to update containerd to a newer version.

---

It works for me a while ago. But today I found that rpull pulls all the layers. And I confirm it by running in the Docker container provided in `/script/demo`, it works. But it just does not work on my host machine. The host machine I am using is running Ubuntu 20.04. I got containerd from `apt install`, and `containerd -v` gives me 
```
containerd github.com/containerd/containerd 1.3.3-0ubuntu2.3 
```

But then I check the `Dockerfile`, it uses `CONTAINERD_VERSION=v1.5.0-beta.4`, therefore I compile a newer version of containerd and it works!!!

**Comment by ktock** [comment]
@nicktming How is this issue going?
